osu.Desktop/Updater/AppImageUpdateManager.cs

```diff
@@ -0,0 +1,564 @@
+                            }
+                            else
+                            {
+                                State = States.CANCELLED;
+                            }
+                            update(progress, State);
+                            process.Dispose();
+                        });
+                    }
+                }
+                else
+                {
+                    update(1, State);
+                }
+            }
+            /// <inheritdoc cref="ApplyUpdate"/>
+            /// <param name="update">Delegate to handle state changes</param>
+            /// <param name="overwrite">Whether to overwrite the AppImage</param>
+            public Task ApplyUpdateAsync(Action<float, States> update = null, bool overwrite = false)
+            {
+                return Task.Run(() => Instance.ApplyUpdate(update, overwrite));
+            }
+        }
+        private INotificationOverlay notificationOverlay;
+        private string version;
+
+        [BackgroundDependencyLoader]
+        private void load(OsuGameBase game, INotificationOverlay notifications)
+        {
+            notificationOverlay = notifications;
+            version = game.Version;
+        }
+
+        protected override async Task<bool> PerformUpdateCheck() => await checkForUpdateAsync().ConfigureAwait(false);
+
+        private async Task<bool> checkForUpdateAsync(UpdateProgressNotification notification = null)
+        {
+            // should we schedule a retry on completion of this check?
+            bool scheduleRecheck = true;
+
+            try
+            {
+                var releases = new OsuJsonWebRequest<GitHubRelease>("https://api.github.com/repos/ppy/osu/releases/latest");
+
+                await releases.PerformAsync().ConfigureAwait(false);
+
+                var latest = releases.ResponseObject;
+
+                // avoid any discrepancies due to build suffixes for now.
+                // eventually we will want to support release streams and consider these.
+                string currentTagName = version.Split('-').First();
+                string latestTagName = latest.TagName.Split('-').First();
+
+                if (latestTagName != version)
+                {
+                    if (AppImageUpdateTool.State == AppImageUpdateTool.States.COMPLETED)
+                    {
+                        // the user may have dismissed the completion notice, so show it again.
+                        notificationOverlay.Post(new UpdateCompleteNotification(this));
+                        return true;
+                    }
+
+                    if (AppImageUpdateTool.HasUpdates)
+                    {
+                        notificationOverlay.Post(new SimpleNotification
+                        {
+                            Text = $"A newer release of osu! has been found ({currentTagName} → {latestTagName}).\n\n"
+                                    + "Click here to download the new version.",
```

Comment:
Probably want update to be initiated automatically, similar to Squirrel on Windows.

Reply:
While doable I was reading in the [doc](https://docs.appimage.org/packaging-guide/optional/updates.html#recommended-user-experience) ?

> Recommended user experience[](https://docs.appimage.org/packaging-guide/optional/updates.html#recommended-user-experience)
> One advantage of the AppImage format is that it gives full control to application authors over the end user experience. Hence, using AppImage and AppImageUpdate, application authors can implement almost any of the schemes outlined above. In order to maintain a consistent and positive user experience with AppImages and AppImageUpdate, we recommend application authors to follow the following Golden Rules:
> 
> Never download updates without the user’s explicit consent, be it in the form of per-update consent, or, optionally, opt-in consent for automatic updates. Thanks for not killing users’ mobile data plans by downloading stuff without asking
> 
> Respect global flags for “do not check for new versions” and “do not attempt to update”. The user may be running a central updating daemon that manages updates for the whole system, in which case any and all attempts to update the application from within itself should be skipped. We need to define those flags for 1) per-system and 2) per-user configuration and 3) ENV (similar to how the old desktopintegration script was set up not to interfere with appimaged)
> 
> Do not bother the user with updates directly as the first thing when the application is launched. When opening an application for the first time, users should see something meaningful to give a positive impression and show immediately what the application is all about (after all, we are automatically taking a screenshot of what your application shows directly after it has been launched for AppImageHub)
> 
> Ask the user for permission before doing version checks. Many open source users value privacy highly and don’t appreciate the “phone home” aspect of forced version checks, which effectively are a form of tracking

Its a general thing though and thus a different issue.

Reply:
Well, if that's how we have it working on Windows I'm not sure Linux should be any different? cc @peppy